### ISSUE TITLE
[BUG] The combination `tf.function` + shot vector + `diff_method="spsa"`  is not supported.

### ISSUE THREAD

**Issue by dwierichs** [PennyLaneAI/pennylane#3715]
### Expected behavior

The following code should execute but doesn't. This is remarked in the corresponding test which is currently being skipped.

```python
import pennylane as qml
import tensorflow as tf

dev = qml.device("default.qubit", wires=2, shots=(10, (5, 1)))

par_0 = tf.Variable(1.5, dtype=tf.float64)
par_1 = tf.Variable(0.7, dtype=tf.float64)

@tf.function
@qml.qnode(dev, diff_method="spsa", interface="tf", max_diff=2)
def circuit(x, y):
    qml.RX(x, wires=[0])
    qml.RY(y, wires=[1])
    qml.CNOT(wires=[0, 1])
    return qml.expval(qml.PauliZ(0) @ qml.PauliX(1))

with tf.GradientTape() as tape1:
    res = circuit(par_0, par_1)
```
Note that the code executes if one 
- changes the `diff_method`,
- sets the `shots` to `None`,
- removes the `tf.function` decorator

### Actual behavior

Doesn't run.

### Additional information

The traceback is from running the currently skipped test case.

### Tracebacks

```shell
def test_hessian_expval_multiple_params(
        self, dev_name, diff_method, gradient_kwargs, shots, num_copies, decorator, interface
    ):
        """The hessian of a single measurement with multiple params return a tuple of arrays."""
    
        #if interface == "tf" and diff_method == "spsa":
            # TODO: Find out why.
            #pytest.skip("SPSA gradient does not support this particular test case")
    
        dev = qml.device(dev_name, wires=2, shots=shots)
    
        par_0 = tf.Variable(1.5, dtype=tf.float64)
        par_1 = tf.Variable(0.7, dtype=tf.float64)
    
        @decorator
        @qnode(dev, diff_method=diff_method, interface=interface, max_diff=2, **gradient_kwargs)
        def circuit(x, y):
            qml.RX(x, wires=[0])
            qml.RY(y, wires=[1])
            qml.CNOT(wires=[0, 1])
            return qml.expval(qml.PauliZ(0) @ qml.PauliX(1))
    
        with tf.GradientTape() as tape1:
            with tf.GradientTape(persistent=True) as tape2:
>               res = circuit(par_0, par_1)

tests/returntypes/tf/test_tensorflow_autograph_qnode_shot_vector.py:369: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../venvs/dev/lib/python3.10/site-packages/tensorflow/python/util/traceback_utils.py:153: in error_handler
    raise e.with_traceback(filtered_tb) from None
/tmp/__autograph_generated_filedg35ntjt.py:378: in grad_fn
    ag__.if_stmt(ag__.ld(mode) == 'forward', if_body_7, else_body_7, get_state_12, set_state_12, ('len_all_params', 'vjps', 'dy'), 2)
/tmp/__autograph_generated_filedg35ntjt.py:372: in else_body_7
    ag__.if_stmt(ag__.converted_call(ag__.ld(isinstance), (ag__.ld(gradient_fn), ag__.ld(qml).gradients.gradient_transform), None, fscope_4), if_body_6, else_body_6, get_state_11, set_state_11, ('len_all_params', 'vjps', 'dy'), 2)
/tmp/__autograph_generated_filedg35ntjt.py:341: in if_body_6
    ag__.if_stmt(ag__.ld(_n) == ag__.ld(max_diff), if_body_5, else_body_5, get_state_10, set_state_10, ('len_all_params', 'vjps', 'dy'), 2)
/tmp/__autograph_generated_filedg35ntjt.py:334: in else_body_5
    vjps = ag__.converted_call(ag__.ld(processing_fn), (ag__.converted_call(ag__.ld(execute), (ag__.ld(vjp_tapes), ag__.ld(device), ag__.ld(execute_fn), ag__.ld(gradient_fn), ag__.ld(gradient_kwargs)), dict(_n=ag__.ld(_n) + 1, max_diff=ag__.ld(max_diff), mode=ag__.ld(mode)), fscope_4),), dict(nums=ag__.ld(output_sizes)), fscope_4)
/tmp/__autograph_generated_filec691u1w8.py:254: in processing_fn
    ag__.for_stmt(ag__.converted_call(ag__.ld(range), (ag__.converted_call(ag__.ld(len), (ag__.ld(tapes),), None, fscope_1),), None, fscope_1), None, loop_body_1, get_state_7, set_state_7, ('start',), {'iterate_names': 't_idx'})
/tmp/__autograph_generated_filec691u1w8.py:182: in loop_body_1
    vjp_ = ag__.converted_call(ag__.ld(processing_fns)[ag__.ld(t_idx)], (ag__.ld(res_t),), dict(num=ag__.ld(nums)[ag__.ld(t_idx)]), fscope_1)
/tmp/__autograph_generated_filew31hhpfl.py:234: in processing_fn
    jac = ag__.converted_call(ag__.ld(fn), (ag__.ld(results),), None, fscope_2)
/tmp/__autograph_generated_filexky_s5r6.py:536: in processing_fn
    ag__.if_stmt(ag__.not_(ag__.ld(shot_vector)), if_body_9, else_body_9, get_state_14, set_state_14, ('grads_tuple',), 1)
/tmp/__autograph_generated_filexky_s5r6.py:529: in else_body_9
    ag__.for_stmt(ag__.converted_call(ag__.ld(range), (ag__.ld(len_shot_vec),), None, fscope_2), None, loop_body_4, get_state_13, set_state_13, (), {'iterate_names': 'idx'})
/tmp/__autograph_generated_filexky_s5r6.py:524: in loop_body_4
    g_tuple = ag__.converted_call(ag__.ld(_single_shot_batch_result), (ag__.ld(res),), None, fscope_2)
/tmp/__autograph_generated_filexky_s5r6.py:489: in _single_shot_batch_result
    ag__.if_stmt(ag__.ld(num_measurements) == 1, if_body_8, else_body_8, get_state_12, set_state_12, ('do_return_1', 'retval__1', 'res'), 2)
/tmp/__autograph_generated_filexky_s5r6.py:406: in if_body_8
    ag__.if_stmt(ag__.ld(num_trainable_params) == 1, if_body_5, else_body_5, get_state_7, set_state_7, ('do_return_1', 'retval__1'), 2)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

    def else_body_5():
        nonlocal do_return_1, retval__1
        try:
            do_return_1 = True
>           retval__1 = ag__.converted_call(ag__.ld(tuple), ((ag__.converted_call(ag__.ld(qml).math.convert_like, (ag__.ld(g), ag__.ld(res)), None, fscope_1) for g in ag__.ld(grads)),), None, fscope_1)
E           tensorflow.python.framework.errors_impl.OperatorNotAllowedInGraphError: Iterating over a symbolic `tf.Tensor` is not allowed in Graph execution. Use Eager execution or decorate this function with @tf.function.
```

### System information

```shell
Name: PennyLane
Version: 0.29.0.dev0
Summary: PennyLane is a Python quantum machine learning library by Xanadu Inc.
Home-page: https://github.com/XanaduAI/pennylane
Author: 
Author-email: 
License: Apache License 2.0
Location: /home/david/venvs/dev/lib/python3.10/site-packages
Requires: appdirs, autograd, autoray, cachetools, networkx, numpy, pennylane-lightning, requests, retworkx, scipy, semantic-version, toml
Required-by: PennyLane-Lightning, PennyLane-qiskit

Platform info:           Linux-5.15.0-58-generic-x86_64-with-glibc2.35
Python version:          3.10.6
Numpy version:           1.23.5
Scipy version:           1.9.3
Installed devices:
- default.gaussian (PennyLane-0.29.0.dev0)
- default.mixed (PennyLane-0.29.0.dev0)
- default.qubit (PennyLane-0.29.0.dev0)
- default.qubit.autograd (PennyLane-0.29.0.dev0)
- default.qubit.jax (PennyLane-0.29.0.dev0)
- default.qubit.tf (PennyLane-0.29.0.dev0)
- default.qubit.torch (PennyLane-0.29.0.dev0)
- default.qutrit (PennyLane-0.29.0.dev0)
- null.qubit (PennyLane-0.29.0.dev0)
- lightning.qubit (PennyLane-Lightning-0.28.2)
- qiskit.aer (PennyLane-qiskit-0.28.0)
- qiskit.basicaer (PennyLane-qiskit-0.28.0)
- qiskit.ibmq (PennyLane-qiskit-0.28.0)
- qiskit.ibmq.circuit_runner (PennyLane-qiskit-0.28.0)
- qiskit.ibmq.sampler (PennyLane-qiskit-0.28.0)
```

### Existing GitHub issues

- [X] I have searched existing GitHub issues to make sure the issue does not already exist.

**Comment by timmysilv** [comment]
Hey @dwierichs! This is a known behaviour of autograph, and we don't plan on addressing it in PennyLane, which is why the tests are being skipped. If you want to leave this issue open to track this, that's all good, but I'm going to mark it as a won't fix.

**Comment by isaacdevlugt** [comment]
Closing this issue as we try to clean up our backlog :)
